### ISSUE TITLE
add a lib-shdc?

### ISSUE THREAD

**Issue by kevinw** [floooh/sokol-tools#34]
Being able to recompile shaders at run-time is a thing I would like in my sokol projects.

sokol-shdc has done a wonderful job of wrapping the `glslang` and `SPIRV-cross` libs. Would it be feasible to write a C API for a sokol-shdc.dll that, given some shader source, fills the appropriate sokol shader description structs?

**Comment by floooh** [comment]
Yeah I think that should work and wouldn't be too complicated (both as DLL or static link library, or with a bit of tinkering even as a WASM module). It won't be small though (the DLL will probably be between 5..6 MBytes, or in case of a static lib, add that much to the executable).

What's needed is an "API header/source pair" which defines a C library API and maybe move some code from main.cc into a common source used both by the sokol-shdc exe and the library.

Ideally the sokol-shdc executable would use the same library API internally too (I'd prefer this because then the library API is automatically tested by the sokol-shdc executable).

I had something similar in mind but for a different use case: eventually I want to get rid of the precompiled binaries in https://github.com/floooh/sokol-tools-bin and want to have a solution where the bulk of the shader compiler is compiled to WASM and then executed through some WASI-launcher, I think this would also benefit from having the core functionality wrapped in a separate library.

**Comment by rcases** [comment]
Also define callback functions for basic file operations (open, read, write, seek, tell ,close). If not defined, use the standard ones.
More or less as [SDL_RWops](https://wiki.libsdl.org/SDL_RWops?highlight=%28%5CbCategoryStruct%5Cb%29%7C%28CategoryIO%29)

**Comment by RandyGaul** [comment]
Would it be possible to get some guidance on implementing a library for shdc? It seems a bit low priority for @floooh so perhaps I could take a stab at it. I’ve also wanted to compile shaders at runtime so they can be easily hotloaded, and to simplify build steps by not requiring build time shader compilation.

**Comment by floooh** [comment]
There's a couple of fairly simple steps, and one I don't know how to solve :)

- In the first step I would isolate all IO calls into a virtual interface which needs to be passed into the code which does IO, this would include fopen, fclose, fread, fwrite, fseek, fgets, ...? (but maybe on a higher level, for instance fseek is only used in the combination of fseek(f, 0, SEEK_END) + fseek(f, 0, SEEK_SET) to get the file size. The reason for this virtual interface would be to allow 'file IO' from memory buffers instead of directly touching the filesystem... this would be the only place where some experimentation might be required to 'make it feel right'...
- Error reporting/logging should probably also be tackled... not sure yet how
- In the second step, move most of the main function into a separate source and function which becomes the libraries 'main entry point': https://github.com/floooh/sokol-tools/blob/9300010ad8c6ecbf68ed50c00ad2036591f8459c/src/shdc/main.cc#L28-L126
    - this new entry point should take an Args object as input, and it also needs a 'virtual IO object' which abstract away the differences between direct file IO and reading/writing memory buffers
- finally the required build system changes, the build process should generally build a static link library, and the executable would just be the main.cc source file and link with the library

...and now the tricky part: D3D and Metal byte code generation:

- for D3D this is done by attempting to load a d3dcompiler.dll and calling into that - sokol_gfx.h is doing the same btw when HLSL shader code is provided
- for Metal this is done by running the Metal shader compiler as command line tool

Both might not be an option (but maybe also not required?) for sokol-shdc as a library, so maybe this can simply be ignored (since sokol_gfx.h can just take shader source code as input).

If anybody wants to tackle this, I would prefer it as a number of smaller PRs instead of a single big one, for instance:

1. implement the virtual IO interface and replace any direct IO usage with this
2. maybe another PR for the error reporting
3. move most of the main function into a new common entry point function, and change main() to call that function
4. implement the actual split into a library and executable

**Comment by kkukshtel** [comment]
Correct me if I'm wrong but bytecode generation would be the thing that would make the library most useful. Especially in terms of hotloading shaders, there would need to be some compile step as part of library functionality to actually consume the shaders, otherwise the library would only be about producing .c versions of the shaders, correct?

For me, the bytecode generation bit would definitely be the thing worth doing. Though I'm also not good enough (for now!) to actually build this func and would mostly be a consumer of whatever anyone else does until I could contribute to it.

**Comment by rcases** [comment]
I've been creating my own library for a while now.

The way I've done it revolves around making minimal changes to the shdc codebase to make it easier to maintain.
These changes are splitting the load_and_preprocess method into load and preprocess and adding parse_source.

The library can use either a filename or a pointer to the source code. 
Unfortunately I haven't yet implemented a way to load includes from a virtual interface.

If you're interested in seeing the strategy I've used, you can find it here:

https://gist.github.com/rcases/8f65e7d54b2556fea60bc138af6cdb0d

**Comment by floooh** [comment]
> otherwise the library would only be about producing .c versions of the shaders, correct?

No, you would get GLSL, HLSL, MSL and WGSL source code as strings which you can feed into sokol-gfx in the `sg_make_shader()` call. The compilation of HLSL and MSL to byte code happens inside sokol_gfx.h then (and this compilation needs to happen either way, because SPIRVCross cannot generate HLSL or Metal bytecode directly).

How to deal with reflection is indeed an interesting question though. One 'naive' way would be to have a new function which directly returns a populated `sg_shader_desc` struct, but that would make sokol-shdc dependent on sokol_gfx.h, which really isn't a good idea.

A cleaner way would be to have a new 'user-friendly' reflection struct which makes it easy to build a `reflection_to_shader_desc()` helper function on the user side.

**Comment by kkukshtel** [comment]
Ah thanks for the clarification!

> but that would make sokol-shdc dependent on sokol_gfx.h, which really isn't a good idea.

One possible option here could be to provide a header similar to sokol_glue.h that acts as a go-between when both gfx and the (not yet existent) sokol_shdc.h are present that can return a populated `sg_shader_desc_ struct?

**Comment by MrCheeseWasTaken** [comment]
I don't know if this is still in the works, but I decided to take a crack at it as I wanted to be able to do this for a project

[Runtime Compiler](https://github.com/MrCheeseWasTaken/sokol-tools-Runtime-Compiler)

I have gotten it to a point where it works... -ish

Sometimes when i run it, it works as expected and I'm able to create a shader at runtime. However other times a warning gets printed:
```
[sg][warning][id:10][line:11114]
[sg][info][id:10][line:11115]
```
which leads to this in sokol_gfx.h:
```
_SG_WARN(GL_UNIFORMBLOCK_NAME_NOT_FOUND_IN_SHADER);
_SG_LOGMSG(GL_UNIFORMBLOCK_NAME_NOT_FOUND_IN_SHADER, u_desc->glsl_name);
```
[(example code)](https://github.com/MrCheeseWasTaken/sokol-tools-Runtime-Compiler/blob/master/src/compiler_example/example.cpp)

I don't know what causes the uniform blocks to sometimes not be found. I have tried checking using renderdoc but the shader seems the exact same as to when it does work.

I have tried to document my changes to sokol-shdc [here](https://github.com/MrCheeseWasTaken/sokol-tools-Runtime-Compiler/blob/master/ShaderCompiler_Changes.md) so if anyone wants to help me out, please do.

I have tried to make the library so that main.cc can still work as is. I had to change some of the functions in the other files to decouple them from using args.cc. The only thing to add is getting the tmpdir for bytecode compiling for metal

Also, I think I'm compiling sokol-shdc wrong... The smallest I managed to get the exe was 15mb, how do I get to it to be smaller?

**Comment by floooh** [comment]
>  don't know what causes the uniform blocks to sometimes not be found. 

The most likely reason for this is dead-code elimination in the GL driver, e.g. GL drivers are free to drop uniforms that are not referenced by code (IIRC this was the reason why this specific log message was added).

I don't know why it would happen sometimes and sometimes not.

> The smallest I managed to get the exe was 15mb

It's roughly in the same ballpark as the 'official' executables (those are between 7.5 and 11 MB, but only after some tinkering with compile and link options):

https://github.com/floooh/sokol-tools/blob/6ad7be69e7b99dfb9b9077b740348ad66b20c036/fibs.ts#L31-L51

**Comment by rcases** [comment]
> The smallest I managed to get the exe was 15mb

I use sokol-shdc as a dll and the 32-bit release version with /MT; it takes up 5.9 MB.

**Comment by MrCheeseWasTaken** [comment]
Back after a few days.

Turns out it was just dangling ```.c_str()```'s that were causing the problem 😅

Everything is working now and everything sokol_shdc can do, the runtime compiler library should also be able to do, though I have not done any tests on this.

Though I don't think the structure of the lib is quite where I want it to be and it is a bit odd to use and not quite 'sokol like' but it works for now. I think @rcases's method would be more in line with it.

**Comment by neslib** [comment]
Like @kkukshtel a DLL version of shdc would be nice for any target language that is not directly supported by Sokol. For example, I created Sokol bindings for Delphi (https://github.com/neslib/Neslib.Sokol) and created a custom version of shdc back then that could output Delphi source code.

To support more language bindings, having a DLL that outputs the converted code and metadata/reflection data would be great. Ideally, it could output this data not only to disk, but also to memory. The virtual file abstraction layer mentioned earlier would be suited for this.

Another option would be to have some sort of plugin system, where binding creators could add a plugin to shdc to output source code in their programming language. 

Since sdhc can already output most if not all required information, I can write a tool that calls sdhc and uses the YAML and shader output to create the Delphi code. But a DLL version would be easier to use.

**Comment by floooh** [comment]
Noted, I'll put the DLL option back into the todo list. The main work is really just to create a common entry-point function which takes an options-struct instead of parsing command line args.

> ...to have some sort of plugin system...

Too complex IMHO, I'd rather extend the YAML output or add a new output option which creates all required files for another tool to take over.

In general more 'runtime modularity' would be good though, but I'd probably rather do that in UNIX fashion by 'pipelining' executables instead of adding the concept of plugin DLLs. E.g. maybe at one point it might make sense to implement the code generators in separate small cmdline tools instead of baking them all into a single exe. And for modularized inputs that problem might come up later when considering to add new language 'frontends' (for instance at some point it would be nice plug slang in front of sokol-shdc (https://github.com/shader-slang/slang)).

(OTH: if we want to let apps integrate sokol-shdc for runtime shader compilation, it must be possible to integrate those modular mini-exes as DLLs anyway hmm... it's actually two different requirements: (1) allow people to extend sokol-shdc without having to provide a PR, and (2) allow people to integrate sokol-shdc into their app as runtime shader compiler).

**Comment by floooh** [comment]
I made a first step in https://github.com/floooh/sokol-tools/commit/aa4f2e315881909a05a996b5b246201ae8705ec8, this basically splits out the core stuff from `main.cc` into a common `shdc.cc`, and the new `Shdc` structs gets a `int run(const Args& args)` function which is then called from main().

The build system now creates a static library with `Shdc::run()` as new 'entry point', and an exe and dll which both link against this common static library.

This is only halfway there though:

- the DLL still has a C++ ABI interface, there is no C API wrapper
- all IO is still file based

Ideally this new Shdc interface should not expose 'internal types' like `Args`, instead if should probably have one function which takes argc+argv, and a separate function which takes a new Options struct. The entire Shdc interface should also probably be a C API in the first place. 

This Options struct should probably also have IO callbacks for opening, reading, writing, closing a 'file' (which can then be overridden by the user to load from and store into memory instead).

**Comment by neslib** [comment]
Thanks! Off the top of my mind, a library version should offer the following functionality:

- a C API for easy integration into other programming languages.
- Provide a function that returns an Options struct populated from the command line. It should probably allow for unknown command line options, so some language backend can add their own command line options as needed. Maybe the Option struct can contain an array with any additional options (as key/value pairs).
- Provide a structure with callbacks for implementing a virtual file interface.
- Provide a function that compiles the shader code and returns the compiled shader variants and the (meta)data needed to polulate the various Sokol desc structs. That function would take the Options struct as parameter to configure the compilation and output.

The metadata could be supplied as YAML output (as the tool currently does already) or it could be a struct that mimics the corresponding structs in Sokol (to avoid a dependency on Sokol itself). It could even be a flat array of fully qualified key/value pairs or something else that's relative easy for you to generate.

You already mentioned some of these solutions earlier. The other ones are just some of my thoughts. Feel free to use the suggestions or not. Anything solution that generates the output required to generate source code in other languages would be great! The only hard requirement would be a flat C API.
